### ISSUE TITLE
Logstash Core Event Java Bug

### ISSUE THREAD

**Issue by driverpt** [elastic/logstash#5137]
I detected a bug on logstash-core-event-java.

We have Pure Java-libraries performing all the Filtering and we do a small "hack" to have @metadata inside the event:

`@some_filter.filter event.to_hash, event["@metadata"]`

And within the Java Filter object, we just put the metadata map inside the event so that we have a Map for everything

`
private static final String METADATA_FIELDREF = "@metadata";

@Override
public void filter(Map<String, Object> eventData, Map<String, Object> metaData) {
  eventData.put(METADATA_FIELDREF, metaData);
  doFilter(eventData);
  eventData.remove(METADATA_FIELDREF);
}
`

This behaviour occurs only on logstash-core-event-java, and this got fixed by switching back to logstash-core-event library.

**Comment by driverpt** [comment]
**UPDATE**

From my analysis, a copy of the object is being made by Rubyfier class at https://github.com/elastic/logstash/blob/master/logstash-core-event-java/src/main/java/com/logstash/Rubyfier.java#L55 .

This is totally not backwards-compatibility, because on logstash-core-event if we do a `to_hash` and we perform operations on the Hash/Map, it will reflect on the original event object.

**Comment by driverpt** [comment]
Why are you using a Rubyfier and copying Java Map to a new RubyHash when JRuby already does that for you?

Source: https://github.com/jruby/jruby/blob/1.7.22/core/src/main/java/org/jruby/java/proxies/MapJavaProxy.java

I understand it makes sense that the Ruby Code see's the Ruby Hash, but since JRuby is the Logstash base, i would not be shocked to see a Java::Util::Map then i call `obj.class`

**Comment by colinsurprenant** [comment]
@driverpt thanks for the report.

The *current* Ruby Event behaviour of handing a reference to its inner data Hash that has the side effect of being in-place mutable (mutating the returned reference mutates the original Event) is a non-hygienic implementation side-effect which is not a good idea to support in the future. Moving forward, this will not be supported and we'll also work on providing better Event *API* documentation.

**Comment by colinsurprenant** [comment]
@driverpt in short the Rubifier is required because some conversions are not covered in the JRuby proxy. Also, we are improving this conversion mechanism, see #4834 and #4999

**Comment by guyboertje** [comment]
@driverpt - If I were you, I would not be using Maps derived from the JRuby Extension.
I would experiment by adding a filter method that takes the Java Event directly and use the public methods in the Java Event to do what ever the filter is doing. The JRuby Event that is passed to your filter should 'see' the changes as it is passed on to the next filter/output.
```java
@Override
public void filter(Event event) {
  doFilter(event);
}
```
```ruby
@some_filter.filter event.getEvent
```
NOTE: We have not tried this ^ yet as we have not yet begun the process of creating pure Java plugins.

You are hereby advised that you are way beyond the bleeding edge with what you are doing and we will try to help but it will not be a high priority for us.

**Comment by driverpt** [comment]
@colinsurprenant , so basically what you're saying is that we should only use Ruby Libraries(Gems) to do all the filtering processing?

@guyboertje , We would like to use the Java Event, but there is no logstash-core-event-java published in Maven Central. (There is a pending PR #4947, but i haven't yet had the time to update it)

But nevertheless, this was the closest solution we could come up with, we have very high performance gains with Java Libraries that are crucial to our needs.

We appreciate all the help and support that you're providing us.

**Comment by guyboertje** [comment]
@driverpt - any chance you could use the jar until it is in Maven or such

**Comment by colinsurprenant** [comment]
@driverpt not sure how/where you read that? It is actually the opposite, all the work we do on the Java Event implementation and future Java pipeline implementation is to move toward providing support for pure Java plugin implementations, or any other JVM languages for that respect, by providing a pure Java API. This is done in an iterative way and we haven't yet defined how a Java plugins could/should be written *today* by leveraging the Java Event (for this #5032 might be interesting). In the short term is it not really possible to completely avoid all Ruby but you can certainly go a long way in having most of your filter logic in Java.

So, basically, you are a bit ahead of the curve here :P but this is also a good opportunity to work with you and see what work or not so please do not hesitate to ask questions, we will help as much as we can but do not expect immediate answers, as @guyboertje noted.

For your Maven request, I will followup shortly in that issue.

**Comment by mjscosta** [comment]
> @driverpt thanks for the report.
> 
> The current Ruby Event behavior of handing a reference to its inner data Hash that has the side effect of being in-place mutable (mutating the returned reference mutates the original Event) is a non-hygienic implementation side-effect which is not a good idea to support in the future. Moving forward, this will not be supported and we'll also work on providing better Event API documentation.

Hi, @colinsurprenant, regarding the data access, I understand your non-hygienic concern, but I'm not sure I agree with it, due to the following reasons.

1- having to_hash, return a copy, duplicates the data, and degrades performance, which may be missused by plug-in developers or even in config file code. Considering the logstash, objective  as a filtering pipeline, the least objects you create the least you degrade performance. I would even say that, you have a model of O(Kn), where K, can be considered a non neglectable constant, where for each cost  you add to processing an event, you degrade the performance by the same amount, almost directly ...

2 - Event with Accessors, how do you address, the returning of nodes of an Event, that have multiple fields ? 
E.g. Event = { a => {b1=>"b1 value", b2 => "b2 value", ...}, c => "c value" }, event["[a]"] = ?
I don't know if this is a common scenario, but if it is, you don't want to make more copies, and you may want to compute, and change values on the node only, and use it as a context, instead of the all event ...

In a sense, I think, you could have a double view of the event, one as a tree of maps (Map<..>), and another using accessors. For which I can imagine different advantages an use cases.

3- Exposing the internal data, does compromises future changes of the event, but provides you with performance, and I don't think they are incompatible. I would see an event more than a tree of maps. And with a framework of services to manipulate the data structure ... 

Sorry for the long writing, but I would just like to give this suggestions.
